flash/data/data_module.py

```diff
@@ -0,0 +1,368 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import platform
+from copy import deepcopy
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+
+import pytorch_lightning as pl
+import torch
+from pytorch_lightning.core.datamodule import _DataModuleWrapper
+from pytorch_lightning.trainer.states import RunningStage
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch.nn import Module
+from torch.utils.data import DataLoader, Dataset
+from torch.utils.data.dataset import Subset
+
+from flash.data.auto_dataset import AutoDataset
+from flash.data.data_pipeline import DataPipeline, Postprocess, Preprocess
+
+
+# TODO: unused?
+class MockLightningModule(pl.LightningModule):
+
+    pass
+
+
+class TaskDataPipeline(DataPipeline):
+
+    def per_batch_transform(self, batch: Any) -> Any:
+        return (batch["x"], batch.get('target', batch.get('y'))) if isinstance(batch, dict) else batch
+
+
+class _FlashDataModuleWrapper(_DataModuleWrapper):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def __call__(self, *args, **kwargs):
+        """A wrapper for ``DataModule`` that:
+
+        TODO: describe what is __flash_special_attr__ for
+        """
+        __flash_special_attr__ = getattr(self, "__flash_special_attr__", None)
+        saved_attr = []
+        if __flash_special_attr__:
+            for special_attr_name in __flash_special_attr__:
+                attr = deepcopy(getattr(self, special_attr_name, None))
+                saved_attr.append((special_attr_name, attr))
+
+        obj = super().__call__(*args, **kwargs)
+
+        if __flash_special_attr__:
+            for special_attr_name, attr in saved_attr:
+                setattr(obj, special_attr_name, attr)
+
+        return obj
+
+
+class DataModule(pl.LightningDataModule, metaclass=_FlashDataModuleWrapper):
+    """Basic DataModule class for all Flash tasks.
+
+    Args:
+        train_ds: Dataset for training.
+        valid_ds: Dataset for validating model performance during training.
+        test_ds: Dataset to test model performance.
+        predict_ds: Dataset for predicting.
+        batch_size: the batch size to be used by the DataLoader.
+        num_workers: The number of workers to use for parallelized loading.
+            Defaults to None which equals the number of available CPU threads,
+            or 0 for MacOS.
+    """
+
+    preprocess_cls = Preprocess
+    postprocess_cls = Postprocess
+
+    def __init__(
+        self,
+        train_ds: Optional[AutoDataset] = None,
```

Comment:
I'd indicate somewhere that this parameters can be a torch Dataset.

Reply:
Change to Dataset.